poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+	}
+	return nil
+}
+
+/* Tests whether two Promise structs are equal
+ *
+ * Arguments
+ *    p2 = a pointer to the struct to test for equality
+ *
+ * Returns
+ *   true if equal, false otherwise
+ */
+func (p *Promise) Equal(p2 *Promise) bool {
+	if p.n != p2.n {
+		return false
+	}
+	if p.shareSuite != p2.shareSuite || p.keySuite != p2.keySuite {
+		return false
+	}
+	for i := 0; i < p.n; i++ {
+		if !p.secrets[i].Equal(p2.secrets[i]) ||
+			!p.insurers[i].Equal(p2.insurers[i]) {
+			return false
+		}
+	}
+	return p.t == p2.t && p.r == p2.r &&
+		p.pubKey.Equal(p2.pubKey) && p.pubPoly.Equal(&p2.pubPoly)
+}
+
+/* Returns the number of bytes used by this struct when marshalled
+ *
+ * Returns
+ *   The marshal size
+ *
+ * Note
+ *   Since the length of insurers and secrets is not known until after
+ *   unmarshalling, do not call before unmarshalling.
+ */
+func (p *Promise) MarshalSize() int {
+	return 3*uint32Size + p.keySuite.PointLen() + p.pubPoly.MarshalSize() +
+		p.n*p.keySuite.PointLen() + p.n*p.shareSuite.SecretLen()
+}
+
+/* Marshals a Promise struct into a byte array
+ *
+ * Returns
+ *   A buffer of the marshalled struct
+ *   The error status of the marshalling (nil if no error)
+ *
+ * Note
+ *   The buffer is formatted as follows:
+ *
+ *      ||n||t||r||pubKey||pubPoly||==insurers_array==||==secrets==||
+ *
+ *   Remember: n == len(insurers) == len(secrets)
+ */
+func (p *Promise) MarshalBinary() ([]byte, error) {
+	buf := make([]byte, p.MarshalSize())
+
+	pointLen := p.keySuite.PointLen()
+	polyLen := p.pubPoly.MarshalSize()
+	secretLen := p.shareSuite.SecretLen()
+
+	// Encode n, r, t
+	binary.LittleEndian.PutUint32(buf, uint32(p.n))
+	binary.LittleEndian.PutUint32(buf[uint32Size:], uint32(p.t))
+	binary.LittleEndian.PutUint32(buf[2*uint32Size:], uint32(p.r))
+
+	// Encode pubKey and pubPoly
+	pointBuf, err := p.pubKey.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(buf[3*uint32Size:], pointBuf)
+
+	polyBuf, err := p.pubPoly.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	copy(buf[3*uint32Size+pointLen:], polyBuf)
+
+	// Encode the insurers and secrets array (Based on poly/sharing.go code)
+	bufPos := 3*uint32Size + pointLen + polyLen
+	for i := range p.insurers {
+		pb, err := p.insurers[i].MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(buf[bufPos+i*pointLen:], pb)
+	}
+	bufPos += p.n * pointLen
+
+	for i := range p.secrets {
+		pb, err := p.secrets[i].MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(buf[bufPos+i*secretLen:], pb)
+	}
+	return buf, nil
+}
+
+/* Unmarshals a Promise from a byte buffer
+ *
+ * Arguments
+ *    buf = the buffer containing the Promise
+ *
+ * Returns
+ *   The error status of the unmarshalling (nil if no error)
+ */
+func (p *Promise) UnmarshalBinary(buf []byte) error {
+	pointLen := p.keySuite.PointLen()
+	secretLen := p.shareSuite.SecretLen()
+
+	// Decode n, r, t
+	p.n = int(binary.LittleEndian.Uint32(buf))
```

Comment:
no need to marshal/unmarshal n, r, t: as per above, they should be well-known config parameters that receiving nodes can be expected to "know" up-front.